### PR TITLE
rgw/s3: add conditional json output for s3:ListBucket

### PR DESCRIPTION
**Pull request by mattbenjamin**
rgw/s3: add conditional json output for s3:ListBucket
    
    Adapt the XML ListBucket response to contain well-formed JSON.
    
    Fixes: https://tracker.ceph.com/issues/53455
    
    Signed-off-by: Matt Benjamin <mbenjamin@redhat.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/rgw/rgw_rest_s3.cc [hunk 1]
```diff
@@ -1809,43 +1809,65 @@ void RGWListBucket_ObjStore_S3::send_response()
...
+      /* conditionally format JSON in the obvious way--I'm unsure if
+       * AWS actually does this */
+      if (s->format == RGWFormat::XML) {
+	s->formatter->open_array_section("Contents");
+	s->formatter->dump_string("Key", key_name);
       } else {
-          s->formatter->dump_string("Type", "Normal");
+	// json
+	s->formatter->open_object_section("dummy");
+	s->formatter->open_object_section(key_name);
```
Comment (cbodley) [inline src/rgw/rgw_rest_s3.cc:1838]: hmm, so the json response will have the outer "Contents" array of unnamed "dummy" objects, and each of those objects contain a single key/value like this?
```json
"Contents": [
  {
    "keyname1": {
      "LastModified": "timestamp",
      "ETag": "etag",
    },
  },
  {
    "keyname2": {
      "LastModified": "timestamp",
      "ETag": "etag",
    },
  },
]
```
i would have expected "Key" to stay at the same level as the rest of the per-object fields like this:
```json
"Contents": [
  {
    "Key": "keyname1",
    "LastModified": "timestamp",
    "ETag": "etag",
  },
  {
    "Key": "keyname2",
    "LastModified": "timestamp",
    "ETag": "etag",
  },
```
that seems like the more direct translation from the xml format

i played around with awscli, and found that it outputs its results in json format by default (although the client is just converting the xml response into json), and this is how it looks:
> $ aws --endpoint-url http://localhost:8000 s3api list-objects --bucket testbucket
```json
{
    "Contents": [
        {
            "Key": "ceph.conf",
            "LastModified": "2022-08-16T16:22:08.109Z",
            "ETag": "\"94697f9200da0f4f5ce3d0959f822275\"",
            "Size": 5371,
            "StorageClass": "STANDARD",
            "Owner": {
                "DisplayName": "M. Tester",
                "ID": "testid"
            }
        },
        {
            "Key": "keyring",
            "LastModified": "2022-08-16T16:21:54.894Z",
            "ETag": "\"7bfa32696bf5a5df785ce3039aefc2ea\"",
            "Size": 294,
            "StorageClass": "STANDARD",
            "Owner": {
                "DisplayName": "M. Tester",
                "ID": "testid"
            }
        }
    ]
}
```

Comment (mattbenjamin) [reply src/rgw/rgw_rest_s3.cc:1838]: ok, thanks for that--it should definitely match what awscli thinks

File: src/rgw/rgw_rest_s3.cc [hunk 2]
```diff
@@ -1809,43 +1809,65 @@ void RGWListBucket_ObjStore_S3::send_response()
...
-        s->formatter->open_array_section("Contents");
-        if (encode_key) {
-          string key_name;
-          url_encode(key.name, key_name);
-          s->formatter->dump_string("Key", key_name);
+  
+  if (op_ret >= 0) {
+    if (s->format == RGWFormat::JSON) {
+      s->formatter->open_array_section("Contents");
+    }
```
Comment (cbodley) [inline src/rgw/rgw_rest_s3.cc:1818]: `optional<Formatter::ArraySection>` might be a good way to conditionally open a section without having to match it with a conditional close. think this helps?
```c++
{
  std::optional<Formatter::ArraySection> json_array;
  if (s->format == RGWFormat::JSON) {
    json_array.emplace("Contents");
  }
  // ...
}
```

Comment (mattbenjamin) [reply src/rgw/rgw_rest_s3.cc:1818]:
```{
  "Name": "works3",
  "Prefix": "",
  "MaxKeys": 1000,
  "IsTruncated": "false",
  "Contents": [
    {
      "Key": "pass1",
      "LastModified": "2022-08-15T17:22:01.609Z",
      "ETag": "\"271ec5227ced4bd1f5de40e6d665b3e1\"",
      "Size": 4458,
      "StorageClass": "STANDARD",
      "Owner": {
        "ID": "testid",
        "DisplayName": "M. Tester"
      },
      "Type": "Normal"
    },
    {
      "Key": "pass2",
      "LastModified": "2022-08-16T12:53:10.667Z",
      "ETag": "\"271ec5227ced4bd1f5de40e6d665b3e1\"",
      "Size": 4458,
      "StorageClass": "STANDARD",
      "Owner": {
        "ID": "testid",
        "DisplayName": "M. Tester"
      },
      "Type": "Normal"
    },
    {
      "Key": "pass3",
      "LastModified": "2022-08-16T12:53:22.206Z",
      "ETag": "\"271ec5227ced4bd1f5de40e6d665b3e1\"",
      "Size": 4458,
      "StorageClass": "STANDARD",
      "Owner": {
        "ID": "testid",
        "DisplayName": "M. Tester"
      },
      "Type": "Normal"
    }
  ],
  "Marker": ""
}```

Comment (mattbenjamin) [reply src/rgw/rgw_rest_s3.cc:1818]: what say you, @cbodley ?

Comment (cbodley) [reply src/rgw/rgw_rest_s3.cc:1818]: it works3!

Comment (mattbenjamin) [reply src/rgw/rgw_rest_s3.cc:1818]: (the boost::optional idiom is cute, but doesn't seem simpler to me)

Comment (mattbenjamin) [reply src/rgw/rgw_rest_s3.cc:1818]: it's a clever overload of raii, maybe I can get more used to that over time

**Review by cbodley — APPROVED** [review]

**Comment by cbodley** [comment]
jenkins test api

**Comment by cbodley** [comment]
jenkins test api

**Comment by cbodley** [comment]
jenkins test api

**Comment by cbodley** [comment]
passed qa in https://pulpito.ceph.com/cbodley-2022-08-18_23:33:15-rgw-wip-cbodley-testing-distro-default-smithi/

**Pull request merged**